### ISSUE TITLE
Can't disable `no_effect`

### ISSUE THREAD

**Issue by OfficialURL** [rust-lang/rust-clippy#7171]
I'm using a `struct` with a custom `DerefMut` implementation that internally updates the object every time it's mutably de-referenced, as a way to check whether it has changed. To force an update, I'm using the following code:

```rust
#[allow(clippy::no_effect)]
&mut *custom_deref;
```

Despite appearances, the second line is not actually a no-op. I can't blame Clippy for not being able to figure that out. However, I'm not able to disable the lint warning, even with the `#[allow(clippy::no_effect)]` line. 

The following code replicates this bug. Note how `&mut *custom_deref;` is not a no-op, and note how the warning still shows despite disabling it.

```rust
struct CustomDeref(i32);

impl std::ops::Deref for CustomDeref {
    type Target = i32;

    fn deref(&self) -> &Self::Target {
        println!("Derefenced!");
        &self.0
    }
}

impl std::ops::DerefMut for CustomDeref {
    fn deref_mut(&mut self) -> &mut Self::Target {
        println!("Mutably dereferenced!");
        &mut self.0
    }
}

fn main() {
    let mut custom_deref = CustomDeref(0);

    #[allow(clippy::no_effect)]
    &mut *custom_deref;
}
```

Output after `cargo clippy`:

```text
warning: statement with no effect
  --> src\main.rs:23:5
   |
23 |     &mut *custom_deref;
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::no_effect)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#no_effect

warning: 1 warning emitted
```

**Comment by camsteffen** [comment]
> I can't blame Clippy for not being able to figure that out.

Actually you can. We can detected custom derefs.
